pages/api/companies.ts

```diff
@@ -1,5 +1,130 @@
+import { PrismaClient } from '@prisma/client';
 import type { NextApiRequest, NextApiResponse } from 'next';
+import { CompanyData, CompanyListData, CompanyPostData } from '../../types/company';
+import { isEmpty, isValidUrl } from '../../utils/strings/validations';
+import { capitalizeEveryWord, removeProtocolAndWwwIfPresent } from '../../utils/strings/formatters';
+import { createJsonResponse, HttpMethod, HttpStatus, rejectHttpMethod } from '../../utils/http/httpHelpers';
+import { withAuth } from '../../utils/auth/jwtHelpers';
+import { ApiResponse, StatusMessageType } from '../../types/apiResponse';
+import { Nullable } from '../../types/utils';
+import { withPrismaErrorHandling } from '../../utils/prisma/prismaHelpers';
 
-export default function handler(req: NextApiRequest, res: NextApiResponse) {
-  res.status(200).end(`Request sent to ${req.url}\n`);
+enum MessageType {
+  COMPANY_ALREADY_EXISTS,
+  COMPANY_CREATED_SUCCESSFULLY,
+  EMPTY_NAME,
+  INVALID_COMPANY_URL,
+  MISSING_NAME,
+  MISSING_COMPANY_URL,
 }
+
+const prisma = new PrismaClient();
+
+const messages = Object.freeze({
+  [MessageType.COMPANY_ALREADY_EXISTS]: { type: StatusMessageType.SUCCESS, message: 'Company already exists.' },
+  [MessageType.COMPANY_CREATED_SUCCESSFULLY]: {
+    type: StatusMessageType.SUCCESS,
+    message: 'Company created successfully.',
+  },
+  [MessageType.EMPTY_NAME]: { type: StatusMessageType.ERROR, message: 'Company name is empty.' },
+  [MessageType.INVALID_COMPANY_URL]: { type: StatusMessageType.ERROR, message: 'Company url is invalid.' },
+  [MessageType.MISSING_NAME]: { type: StatusMessageType.ERROR, message: 'Company name is missing.' },
+  [MessageType.MISSING_COMPANY_URL]: { type: StatusMessageType.ERROR, message: 'Company url is missing.' },
+});
+
+function handler(req: NextApiRequest, res: NextApiResponse) {
+  const method = req.method;
+
+  switch (method) {
+    case HttpMethod.GET:
+      handleGet(req, res);
+      break;
+    case HttpMethod.POST:
+      handlePost(req, res);
+      break;
+    default:
+      rejectHttpMethod(res, method);
+  }
+}
+
+async function handleGet(req: NextApiRequest, res: NextApiResponse<ApiResponse<CompanyListData[]>>) {
+  if (req.body.searchKeyword === undefined || req.body.searchKeyword === null) {
```

Comment:
If the searchKeyword isn't provided, shouldn't we return all companies? I also don't quite get why we are returning [] as the result.

Also, if the searchKeyword is empty or whitespace only, shouldn't we have the same behaviour as undefined searchKeyword?

Reply:
I think we will list companies only when people are trying to select them when creating an application? So they should already have a company in mind. So if they haven't typed anything in the bar, then we don't need to suggest any companies yet. I will update to return nothing for the empty string case.

But do you think we need to return all companies? Why?

Reply:
I agree with @zognin here.